src/Microsoft.OpenApi.Readers/ParsingContext.cs

```diff
@@ -1,30 +1,108 @@
+        private IOpenApiVersionService _versionService;
+        internal RootNode RootNode { get; set; }
+        internal List<OpenApiTag> Tags { get; private set; } = new List<OpenApiTag>();
+
+
+        /// <summary>
+        /// Initiates the parsing process.  Not thread safe and should only be called once on a parsing context
+        /// </summary>
+        /// <param name="yamlDocument"></param>
+        /// <param name="diagnostic"></param>
+        /// <returns>An OpenApiDocument populated based on the passed yamlDocument </returns>
+        internal OpenApiDocument Parse(YamlDocument yamlDocument, OpenApiDiagnostic diagnostic)
+        {
+            RootNode = new RootNode(this, diagnostic, yamlDocument);
+
+            var inputVersion = GetVersion(RootNode);
+
+            OpenApiDocument doc;
+            switch (inputVersion)
+            {
+                case "2.0":
+                    this.ReferenceService = new OpenApiV2VersionService();
+                    doc = this.ReferenceService.LoadOpenApi(RootNode);
+                    break;
+
+                default:
+                    this.ReferenceService = new OpenApiV3VersionService();
+                    doc = this.ReferenceService.LoadOpenApi(RootNode);
+                    break;
+            }
+
+            return doc;
+        }
+
+        /// <summary>
+        /// Gets the version of the Open API document.
+        /// </summary>
+        private static string GetVersion(RootNode rootNode)
+        {
+            var versionNode = rootNode.Find(new JsonPointer("/openapi"));
+
+            if (versionNode != null)
+            {
+                return versionNode.GetScalarValue();
+            }
+
+            versionNode = rootNode.Find(new JsonPointer("/swagger"));
+
+            return versionNode?.GetScalarValue();
+        }
+
+        private void ComputeTags(List<OpenApiTag> tags, Func<MapNode,OpenApiTag> loadTag )
+        {
+            // Precompute the tags array so that each tag reference does not require a new deserialization.
+            var tagListPointer = new JsonPointer("#/tags");
+
+            var tagListNode = RootNode.Find(tagListPointer);
+
+            if (tagListNode != null && tagListNode is ListNode)
+            {
+                var tagListNodeAsListNode = (ListNode)tagListNode;
+                tags.AddRange(tagListNodeAsListNode.CreateList(loadTag));
+            }
+        }
+
 
-        private readonly Dictionary<string, IOpenApiReferenceable> _referenceStore =
-            new Dictionary<string, IOpenApiReferenceable>();
 
-        private readonly Dictionary<string, object> _tempStorage = new Dictionary<string, object>();
 
         /// <summary>
         /// Reference service.
         /// </summary>
-        internal IOpenApiReferenceService ReferenceService { get; set; }
+        internal IOpenApiVersionService ReferenceService
```

Comment:
>ReferenceService [](start = 40, length = 16)

VersionService. You can remove the // summary doc above if you like :)

Reply:
Yeah, I noticed I'd forgot that renaming.

Reply:
Fixed